assignments/burmanm/week-5.scm

```diff
@@ -0,0 +1,583 @@
+
+;; Exercise 2.37
+
+;; From SICP
+(define (dot-product v w)
+  (accumulate + 0 (map * v w)))
+
+(define (matrix-*-vector m v)
+  (map (lambda (x) (dot-product x v)) m))
+
+(define (transpose mat)
+  (accumulate-n
+   cons '() mat))
+
+(define (matrix-*-matrix m n)
+  (let ((cols (transpose n)))
+    (map
+     (lambda (x) (matrix-*-vector cols x))       
+     m)))
+
+;; Exercise 2.38
+
+(define (fold-left op initial sequence)
+  (define (iter result rest)
+    (if (null? rest)
+	result
+	(iter (op result (car rest))
+	      (cdr rest))))
+  (iter initial sequence))
+
+(define fold-right accumulate)
+
+(fold-right / 1 (list 1 2 3))
+(fold-left / 1 (list 1 2 3))
+(fold-right list nil (list 1 2 3))
+(fold-left list nil (list 1 2 3))
+
+;; $121 = 3/2
+;; $122 = 1/6
+;; $123 = (1 (2 (3 ())))
+;; $124 = (((() 1) 2) 3)
+
+;; Any property were the order of the operands makes no difference to execution (obviously). For example, * or + (but not - for example)
+
+;; Exercise 2.39
+
+;; (define (reverse list)
+;;   (if (null? list)
+;;       '()
+;;       (append (reverse (cdr list)) (cons (car list) '()))))
+
+(define (reverse sequence)
+  (fold-right (lambda (x y)
+		(append y (cons x '()))) nil sequence))
+
+(define (reverse sequence)
+  (fold-left (lambda (x y)
+		(append (cons y '()) x)) nil sequence))
+
+;; Test, both return the same..
+
+;; scheme@(guile-user) [55]> (reverse (list 1 2 3))
+;; $130 = (3 2 1)
+
+;; Exercise 2.40
+
+;; 1 <= i < j <= n
+
+(define (enumerate-interval low high)
+  (if (> low high)
+      nil
+      (cons low (enumerate-interval (+ low 1) high))))
+
+(define (flatmap proc seq)
+  (accumulate append nil (map proc seq)))
+
+(define (unique-pairs n)
+  (flatmap (lambda (i)
+	     (map (lambda (j) (list i j))
+		  (enumerate-interval 1 (- i 1))))
+  (enumerate-interval 1 n)))
+
+;; Test
+
+;; scheme@(guile-user) [59]> (unique-pairs 5)
+;; $137 = ((2 1) (3 1) (3 2) (4 1) (4 2) (4 3) (5 1) (5 2) (5 3) (5 4))
+
+;; Simplify prime-sum-pairs..
+
+(define (make-pair-sum pair)
+  (list (car pair) (cadr pair) (+ (car pair) (cadr pair))))
+
+(define (prime-sum-pairs n)
+  (map make-pair-sum
+       (filter prime-sum? (flatmap
+			   (lambda (i)
+			     (map (lambda (j) (list i j))
+				  (enumerate-interval 1 (- i 1))))
+			   (enumerate-interval 1 n)))))
+
+(define (prime-sum? pair)
+  (prime? (+ (car pair) (cadr pair))))
+
+;; Test
+
+;; Original:
+
+;; scheme@(guile-user) [58]> (prime-sum-pairs 5)
+;; $134 = ((2 1 3) (3 2 5) (4 1 5) (4 3 7) (5 2 7))
+
+(define (prime-sum-pairs n)
+  (map make-pair-sum
+       (filter prime-sum? (unique-pairs n))))
+
+;; Modified:
+
+;; scheme@(guile-user) [59]> (prime-sum-pairs 5)
+;; $136 = ((2 1 3) (3 2 5) (4 1 5) (4 3 7) (5 2 7))
+
+
+;; Exercise 2.41
+
+(define (make-triplet-sum triplet)
+  (+ (car triplet) (cadr triplet) (caddr triplet)))
+
+(define (create-triple-permutations n)
+  (flatmap (lambda (i)
+	     (map (lambda (j) (cons i j)) (unique-pairs n)))
+	   (enumerate-interval 1 n)))
```

Comment:
It seems like you can't re-use unique-pairs like so because this generates non unique triple. I was able to make it work by updating unique-pairs to use a starting integer, and then call it with `(unique-pairs (+ i 1) n))`.